### PR TITLE
spec source for new extensions

### PR DESCRIPTION
**Pull request by bashbaug**
This PR contains the spec source for the following new extensions:

* `cl_khr_async_copy_fence` (was provisional, now final)
* `cl_khr_extended_async_copies` (was provisional, now final)
* `cl_khr_expect_assume`
* `cl_khr_command_buffer` (provisional)

### CODE REVIEW

File: ext/cl_khr_command_buffer.asciidoc
Comment (bashbaug) [inline ext/cl_khr_command_buffer.asciidoc:368]: Minor/consistency: Most of the other query enums do not include "INFO" in their names.  For example, we have a query enum `CL_CONTEXT_REFERENCE_COUNT` and not `CL_CONTEXT_INFO_REFERENCE_COUNT`.  Should we drop "INFO" in the names of these enums?

Note that the enum type does typically include "info" in the name, so we should keep `cl_command_buffer_info_khr`.

Comment (bashbaug) [reply ext/cl_khr_command_buffer.asciidoc:368]: Tagging @EwanC for awareness.

**Review by alycm — COMMENTED** [review]
Not reviewed `cl_khr_command_buffer` yet, but reviewed the others.

**Comment by Hopetech** [comment]
Thanks you Ben,
`cl_khr_extended_async_copies` looks good.

File: ext/cl_khr_command_buffer.asciidoc
Comment (EwanC) [reply ext/cl_khr_command_buffer.asciidoc:368]: Ah yes, good catch Ben. Happy for you to remove the `_INFO_` component of these enums.

File: ext/cl_khr_expect_assume.asciidoc
Comment (bashbaug) [reply ext/cl_khr_expect_assume.asciidoc:62]: Good suggestion, it couldn't hurt, especially because this behavior isn't described specifically here.

File: ext/cl_khr_command_buffer.asciidoc
Comment (bashbaug) [reply ext/cl_khr_command_buffer.asciidoc]: I removed "INFO" in commit https://github.com/KhronosGroup/OpenCL-Docs/pull/711/commits/929586586711de05bd45c05840c3b3d147d4e949 - please take a look.

Comment (EwanC) [reply ext/cl_khr_command_buffer.asciidoc]: Looks good, updated my headers PR in https://github.com/KhronosGroup/OpenCL-Headers/pull/182/commits/eba2023490a4071718bad97ffbef6f877f41bb2e to match

File: ext/cl_khr_command_buffer.asciidoc
Comment (alycm) [inline ext/cl_khr_command_buffer.asciidoc:1641]:
```suggestion
| *CL_COMMAND_BUFFER_REFERENCE_COUNT_KHR* footnote:[{fn-reference-count-usage}]
| cl_uint
| Return the _command_buffer_ reference count.
```

Will only work if https://github.com/KhronosGroup/OpenCL-Docs/pull/712 is merged first.

**Review by alycm — COMMENTED** [review]
LGTM, but one area that I think we'll have to take to an issue.

File: ext/cl_khr_command_buffer.asciidoc
Comment (bashbaug) [reply ext/cl_khr_command_buffer.asciidoc:1643]: I think this makes sense, but I'll hold off committing this suggestion until #712 is merged.

File: ext/cl_khr_command_buffer.asciidoc
Comment (bashbaug) [reply ext/cl_khr_command_buffer.asciidoc:508]: I don't think this is wrong per se, though it is a little odd and I agree it is semi-recursive.

FWIW, command queue properties has a similar issue, and it's "solved" by having `cl_command_queue_properties` be the bitfield type specifying the actual property bits and `cl_queue_properties` be the pointer to the property-value pairs passed to `clCreateCommandQueueWithProperties`.  This always seemed a little icky and I can never remember which type gets used for which purpose.  I wonder if we would have been better off with just `cl_command_queue_properties` and a similar semi-recursive usage.

Comment (EwanC) [reply ext/cl_khr_command_buffer.asciidoc:508]: When I based this off of `clCreateCommandQueueWithProperties`, I completely missed that `cl_queue_properties` != `cl_command_queue_properties` 🤦 So the semi-recursive relationship wasn't intentional. I think my preference would be for two separate types but not sure if that's too significant a change at this point.

Comment (alycm) [reply ext/cl_khr_command_buffer.asciidoc:508]: > I don't think this is wrong per se

> I think my preference would be for two separate types but not sure if that's too significant a change at this point.

Stepping back for a moment, `cl_bitfield` and `cl_properties` are all just a `cl_ulong`, so this may be more trivial that I was thinking.

Comment (bashbaug) [reply ext/cl_khr_command_buffer.asciidoc:510]: I think I moved the changes we discussed over correctly, please check https://github.com/KhronosGroup/OpenCL-Docs/pull/711/commits/5dda69bfb85d13f7a86d4f847b6036b70a6b7a96.

Comment (EwanC) [reply ext/cl_khr_command_buffer.asciidoc:510]: change LGTM, updated headers PR to reflect it in https://github.com/KhronosGroup/OpenCL-Headers/pull/182/commits/eafab6a1c77007d30f334db475949cf983009307

**Review by alycm — APPROVED** [review]

**Pull request merged**
